src/modules/ModuleRunner.ts

```diff
@@ -66,13 +82,67 @@ export class ModuleRunner {
         return merged;
     }
 
+    /**
+     * Ensure we register extensions provided by the modules
+     */
+    private updateExtensions(): void {
+        const cryptoSetupExtensions: Array<RuntimeModule> = [];
+        const experimentalExtensions: Array<RuntimeModule> = [];
+
+        this.modules.forEach((m) => {
+            /* Record the cryptoSetup extensions if any */
+            if (m.module.extensions?.cryptoSetup) {
+                cryptoSetupExtensions.push(m.module);
+            }
+
+            /* Record the experimantal extensions if any */
+            if (m.module.extensions?.experimental) {
+                experimentalExtensions.push(m.module);
+            }
+        });
+
+        /* Enforce rule that only a single module may provide a given extension */
+        if (cryptoSetupExtensions.length > 1) {
+            throw new Error(
+                `cryptoSetup extension is provided by modules ${cryptoSetupExtensions
+                    .map((m) => m.moduleName)
+                    .join(", ")}, but can only be provided by a single module`,
+            );
+        }
+        if (experimentalExtensions.length > 1) {
+            throw new Error(
+                `experimental extension is provided by modules ${experimentalExtensions
+                    .map((m) => m.moduleName)
+                    .join(", ")}, but can only be provided by a single module`,
+            );
+        }
+
+        /* Override the default extension if extension was provided by a module */
+        if (cryptoSetupExtensions.length == 1) {
+            this.extensions.cryptoSetup = cryptoSetupExtensions[0].extensions?.cryptoSetup;
+        }
+
+        if (experimentalExtensions.length == 1) {
+            this.extensions.experimental = cryptoSetupExtensions[0].extensions?.experimental;
+        }
```

Comment:
Why do we replace individual properties within `AllExtensions` here, whilst in `reset` we replace the entire `AllExtensions` object? It seems inconsistent.

It might be better to treat `AllExtensions` as immutable?

(It might be better still if the definition of `AllExtensions` enforced that, by making the individual properties `readonly`, but since that's part of the module API, it would be a separate PR. Let's at least be consistent here.)

Reply:
I agree that encapsulation could be better. Part of the reason for the design is that any given type of extension, can be provided by any of the registered modules. We really don't know if we have multiple, a single, or no implementations of an extension until we have registered all. 

Regardless, based on the need for better encapsulation, I have refactored the code and introduced an Extension manager which in effect makes the exposed extensions immutable. This provides the proper encapsulation (bar the reset method, which is clearly marked for  test only). 

The refactor right now has the ExtensionManager living in matrix-react-sdk, and is therefore compatible with the current version (2.4.0) of matrix-react-sdk-module-api package. The plan is to move the ExtensionManager to the matrix-react-sdk-module-api package in a new PR. 

This means ModuleRunner will only need to know about (import) the ExtensionsManager, and not the individual extension interfaces or their default implementations.

Reply:
I'm afraid I'm not really following this.

> Part of the reason for the design is that any given type of extension, can be provided by any of the registered modules. We really don't know if we have multiple, a single, or no implementations of an extension until we have registered all.

Why does this preclude making `AllExtensions` immutable?

But ok, it looks like you're moving functionality from `ModuleRunner.updateExtensions` to `ExtensionsManager`, which is fine, though I'm not sure it was necessary. Is the intention that `AllExtensions` be removed once `ExtensionsManager` is added to the module-api package?